New syncing logic

twitu opened the pull request:
Closes #37

kirillt commented:
I think we have also discussed "smart rounding":
> round HH:MM:ss:xxx always to the lowest side when we memoize the timestamp of latest scan
> round HH:MM:ss:yyy always to the highest side when we retrieve the timestamp of file modification

twitu commented:
SystemTime records time at a nanosecond level depending on the operating system. The `as_millis` function implements the standard logic for rounding the timestamp up to milliseconds. The default implementation should be sufficient for our purpose.

```rust
    pub const fn as_millis(&self) -> u128 {
        self.secs as u128 * MILLIS_PER_SEC as u128 + (self.nanos.0 / NANOS_PER_MILLI) as u128
    }
```

twitu commented:
If we want to reduce the possibility of a race condition, we can use nano second resolution without any loss in terms of performance and increase in complexity. We are already is storing the SystemTime which is represented as a `u128`.

```rust
    pub const fn as_nanos(&self) -> u128 {
        self.secs as u128 * NANOS_PER_SEC as u128 + self.nanos.0 as u128
    }
```

kirillt commented:
@twitu yes, nanos are possible and are provided by the filesystem API, but iirc they are not real nanos and there are zeros at the end.. But why you don't want to adopt different rounding logic for reading timestamp? I think this would solve the issue for any precision.

twitu commented:
I just noticed that because of the recent changes in the monoids PR. The semantics for the syncing logic has changed. We're now checking for timestamp equality, so rounding logic won't work. We can only increase the resolution of the timestamp.

```rust
    /// Compare the timestamp of the storage file
    /// with the timestamp of the in-memory storage update
    /// to determine if either of the two requires syncing.
    fn needs_syncing(&self) -> Result<bool> {
        match fs::metadata(&self.path) {
            Ok(metadata) => {
                let get_duration_since_epoch = |time: SystemTime| {
                    time.duration_since(SystemTime::UNIX_EPOCH)
                        .unwrap()
                        .as_secs()
                };

                let fs_modified =
                    get_duration_since_epoch(metadata.modified()?);
                let self_modified = get_duration_since_epoch(self.modified);

                Ok(fs_modified != self_modified)
```

kirillt commented:
> I just noticed that because of the recent changes in the monoids PR. The semantics for the syncing logic has changed. We're now checking for timestamp equality, so rounding logic won't work. We can only increase the resolution of the timestamp.

Agree, this changes everything...

Ready to merge as soon as build is fixed.

twitu commented:
A completely revamped syncing logic. This affects `read_fs`, `write_fs`, `new` FileStorage and some minor changes to the `BaseStorage` api.

Needs a solid review.

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -176,9 +157,77 @@ where
+        + Monoid<V>,
+{
+    /// Set a key-value pair in the storage
+    fn set(&mut self, key: K, value: V) {
+        self.data.entries.insert(key, value);
+        self.modified = std::time::SystemTime::now();
+    }
```
Why not write to the file system whenever a user sets a value?

kirillt replied:
We've discussed this in the other PR. The idea is to allow the user to control when exactly the write is happening. Performance reasons mostly. Storage values can be updated often, but writing to disk is a slow operation. Also, we run coroutines in Kotlin to update storages. It's difficult to do sequentially because it's IO and we don't want to freeze UI thread. So, if we create lots of coroutines they start race with each other. We have atomic versions exactly for this reason, but it's also nice to just reduce the load.

tareknaser replied:
Thanks for clarifying. I think I had forgotten about this.

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -176,9 +157,77 @@ where
+    }
+
+    /// Remove a key-value pair from the storage given a key
+    fn remove(&mut self, id: &K) -> Result<()> {
+        self.data.entries.remove(id).ok_or_else(|| {
+            ArklibError::Storage(self.label.clone(), "Key not found".to_owned())
+        })?;
+        self.modified = std::time::SystemTime::now();
+        Ok(())
+    }
```
Here too. 

If a user removes a value from `FileStorage` instance, I think we should update file system to reflect the changes. 
What do you think?

kirillt replied:
Explained the decision in comment to `fn set`.

tareknaser commented on fs-storage/src/base_storage.rs:
```diff
@@ -1,27 +1,37 @@
+}
+
+/// A trait for a key-value mapping that is persisted to the file system.
 pub trait BaseStorage<K, V>: AsRef<BTreeMap<K, V>> {
```
I think we should actually start here. The `BaseStorage` trait's methods should have explicit documentation comments explaining their functionality. Questions such as whether to write to the file system when setting/removing a value, when to sync, and when to call `merge_from` should be clearly stated in the trait itself, as these methods work together. 

The implementation in `FileStorage` should only contain the logic defined in the trait.

twitu replied:
I've updated the comments. I've made it clear that the BaseStorage trait only updates the mapping and it is the responsibilty of the implementor to decide when to interact with the file system.

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -176,9 +157,77 @@ where
+    /// with the timestamp of the in-memory storage and the last written
+    /// to time to determine if either of the two requires syncing.
+    fn needs_syncing(&self) -> Result<SyncStatus> {
+        let file_updated = fs::metadata(&self.path)?.modified()?;
+
+        // mapping updated
+        if self.modified > self.written_to_disk {
+            // file updated since last write
+            if file_updated > self.written_to_disk {
+                // both updated
+                return Ok(SyncStatus::FullSync);
+            } else {
+                // only mapping updated
+                return Ok(SyncStatus::DownSync);
+            }
+        } else {
+            // file updated since last write
+            if file_updated > self.written_to_disk {
+                // only file updated
+                return Ok(SyncStatus::UpSync);
+            } else {
+                // no change
+                return Ok(SyncStatus::NoSync);
+            }
```
I'm sorry, but could you explain the logic here a bit more? I'm not sure I get it.

kirillt replied:
Actually, previous `match` expression looked more readable to me.

@tareknaser Ishan introduced one more timestamp, so now we track times of both internal modification and writing to the filesystem. We have third timestamp in the filesystem itself. These 3 timestamps are enough to figure out which side updated, or if both sides been updated.

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -62,75 +72,43 @@ where
             label,
             path: PathBuf::from(path),
-            modified: SystemTime::now(),
+            modified: time,
+            written_to_disk: time,
```
When do these 2 fields have different values?

twitu replied:
When a value is set only `modified` is updated to current time. So it becomes `modified > written_to_disk` which shows that mapping is ahead of file.

When `write_fs` or `read_fs` is called it becomes `modified = written_to_disk` to reflect that mapping and file are in sync.

kirillt replied:
I think it's worth to document these invariants as comments to `modified` and `written_to_disk` variables.

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -299,34 +355,35 @@ mod tests {
         std::thread::sleep(std::time::Duration::from_secs(1));
         file_storage.set("key1".to_string(), "value1".to_string());
-        assert_eq!(file_storage.needs_syncing().unwrap(), true);
+        assert_eq!(file_storage.needs_syncing().unwrap(), SyncStatus::DownSync);
```
Thank you for adding the tests. 

Could you include additional tests for the `SyncStatus::UpSync` and `SyncStatus::FullSync` cases?

tareknaser commented on fs-storage/src/file_storage.rs:
The `test_file_storage_is_storage_updated` test is failing on macOS. Could you take a look at it? 

Here is the output:

```sh
---- file_storage::tests::test_file_storage_is_storage_updated stdout ----
thread 'file_storage::tests::test_file_storage_is_storage_updated' panicked at fs-storage/src/file_storage.rs:360:9:
error: test failed, to rerun pass `-p fs-storage --lib`
assertion `left == right` failed
  left: UpSync
 right: NoSync
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

twitu replied:
Yes there is some race condition happening there. It's good that the tests have caught it. I'll fix the syncing logic.

tareknaser reviewed (commented):
Thank you for putting this together. I've left some comments. 

There are also a few clippy warnings causing the CI to fail. I've reviewed them, and they're straightforward to address. 
You can run this command to see them
```sh
cargo clippy --workspace --bins -- -D warnings
```

kirillt commented on fs-storage/src/base_storage.rs:
```diff
@@ -1,27 +1,37 @@
 use std::collections::BTreeMap;
 
+#[derive(Debug, PartialEq, PartialOrd, Ord, Eq, Clone)]
+/// Represents the synchronization status of the storage.
+pub enum SyncStatus {
+    /// No synchronization needed.
+    NoSync,
+    /// Load key-value mapping from the file system.
+    UpSync,
+    /// Write key-value mapping to the file system.
+    DownSync,
+    /// Changes from both in-memory and file system need to be merged before syncing.
+    FullSync,
+}
```
Could you describe the motivation behind this separation? My guess is that `DownSync` is what should be used with cache storages, and any other storages seem to require `FullSync`? But when do we need `UpSync` or `NoSync`?

kirillt commented on fs-storage/src/base_storage.rs:
```diff
@@ -1,27 +1,37 @@
 use std::collections::BTreeMap;
 
+#[derive(Debug, PartialEq, PartialOrd, Ord, Eq, Clone)]
+/// Represents the synchronization status of the storage.
+pub enum SyncStatus {
+    /// No synchronization needed.
+    NoSync,
+    /// Load key-value mapping from the file system.
+    UpSync,
+    /// Write key-value mapping to the file system.
+    DownSync,
+    /// Changes from both in-memory and file system need to be merged before syncing.
+    FullSync,
+}
```
Probably I would reword names of these values like "InSync", "Ours", "Theirs", "Both" to accent on which side should be updated. This would also make the enum potentially re-usable with other kinds of backend (filesystem is only starting point, later we might use Iroh directly).

twitu replied:
```
    /// No synchronization needed.
    InSync,
    /// In-memory key-value mapping is stale.
    MappingStale,
    /// External file system storage is stale.
    StorageStale,
    /// In-memory key-value mapping and external file system storage diverge.
    Diverge,
```

How about this? Ours and theirs can become confusing. Since this dealing with FileStorage, mapping and storage makes it clear what we're refering to.

kirillt replied:
This looks good. Ours/theirs is kinda conventional but I agree that it can be confusing.

kirillt commented on fs-storage/src/file_storage.rs:
```diff
@@ -100,37 +109,29 @@ where
+            (true, true, _) => Ok(SyncStatus::DownSync),
+            (true, false, _) => Ok(SyncStatus::FullSync),
+            (_, _, true) => Ok(SyncStatus::UpSync),
+            _ => Ok(SyncStatus::NoSync),
```
This looks cool 👍

kirillt replied:
This looks cool 👍

twitu commented on fs-storage/src/file_storage.rs:
```diff
@@ -62,75 +72,43 @@ where
-            Err(e) => {
-                Err(ArklibError::Storage(self.label.clone(), e.to_string()))
-            }
+    pub fn sync(&mut self) -> Result<()> {
+        match self.needs_syncing()? {
+            SyncStatus::NoSync => Ok(()),
+            SyncStatus::DownSync => self.read_fs().map(|_| ()),
+            SyncStatus::UpSync => self.write_fs(),
+            SyncStatus::FullSync => self.full_sync(),
```
I've added the log statement to the inner function - `sync_status` since it might be called in other places as well.

tareknaser replied:
Nice. Thank you

twitu commented on fs-storage/src/file_storage.rs:
```diff
@@ -62,75 +72,43 @@ where
             },
+        };
+
+        if Path::exists(path) {
+            let _ = storage.read_fs();
         }
```
Fixed.

twitu commented on fs-storage/src/file_storage.rs:
I've added `sync` to `BaseStorage` trait and added doc comments. I've inlined full_sync since it's reusing BaseStorage methods and FileStorage specific.

kirillt commented on fs-storage/src/file_storage.rs:
```diff
@@ -62,75 +72,28 @@ where
     /// Create a new file storage with a diagnostic label and file path
-    pub fn new(label: String, path: &Path) -> Self {
-        Self {
+    pub fn new(label: String, path: &Path) -> Result<Self> {
```
Also worth to mention in the comment that it's initialized with the data stored by the path, if the path exists.

kirillt commented on fs-storage/src/file_storage.rs:
```diff
@@ -176,12 +142,100 @@ where
+        + std::str::FromStr
+        + Monoid<V>,
+{
+    /// Set a key-value pair in the storage
```
```suggestion
    /// Set a key-value pair in the internal mapping
```

kirillt commented on fs-storage/src/file_storage.rs:
```diff
@@ -176,12 +142,100 @@ where
+        self.modified = std::time::SystemTime::now();
+    }
+
+    /// Remove a key-value pair from the storage given a key
```
```suggestion
    /// Remove an entry from the internal mapping given a key
```

kirillt commented on fs-storage/src/file_storage.rs:
```diff
@@ -62,75 +72,28 @@ where
         + Monoid<V>,
 {
     /// Create a new file storage with a diagnostic label and file path
```
```suggestion
    /// Create a new file storage with a diagnostic label and file path
    /// The storage will be initialized using the disk data, if the path exists
```

kirillt commented on fs-storage/src/file_storage.rs:
```diff
@@ -176,12 +142,100 @@ where
+        Ok(status)
+    }
+
+    /// Sync the in-memory storage with the storage on disk
+    fn sync(&mut self) -> Result<()> {
+        match self.sync_status()? {
+            SyncStatus::InSync => Ok(()),
+            SyncStatus::MappingStale => self.read_fs().map(|_| ()),
+            SyncStatus::StorageStale => self.write_fs(),
+            SyncStatus::Diverge => {
+                let data = self.load_fs_data()?;
+                self.merge_from(&data)?;
+                self.write_fs()?;
+                Ok(())
+            }
+        }
+    }
```
👍

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -62,75 +78,29 @@ where
     /// Create a new file storage with a diagnostic label and file path
-    pub fn new(label: String, path: &Path) -> Self {
-        Self {
+    /// The storage will be initialized using the disk data, if the path exists
```
```suggestion
    /// The storage will be initialized using the disk data, if the path exists
    ///
    /// Note: if the file storage already exists, the data will be read from the file
    /// without overwriting it.
```

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -176,12 +149,100 @@ where
+        let status = if self.modified > self.written_to_disk {
+            // file updated since last write
+            if file_updated > self.written_to_disk {
+                // both updated
```
We are only here if the user manually modifies the storage file on disk, correct?

twitu replied:
Externally modified yes. Either by the user or suppose by a mirror storage (like in the test).

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -292,41 +361,73 @@ mod tests {
     }
 
+    #[test]
+    fn test_file_timestamp_bug() {
```
Could we remove this test case now that we've fixed the bug?

tareknaser reviewed (commented):
Thanks for all your work on this! I think we're almost ready to merge. 

Could you also rebase it on top of main?

twitu commented:
Another weird timestamp related bug. It seems `write_fs` doesn't even change the modified timestamp in the file metadata.

```
before_write: SystemTime { tv_sec: 1718114429, tv_nsec: 471980190 }, after_write: SystemTime { tv_sec: 1718114429, tv_nsec: 471980190 }
thread 'file_storage::tests::test_file_storage_is_storage_updated' panicked at fs-storage/src/file_storage.rs:406:9:
```

tareknaser commented:
> Another weird timestamp related bug. It seems `write_fs` doesn't even change the modified timestamp in the file metadata.

Could you write a minimal test to reproduce it? 
locally for me, all tests are passing at 4a13396

twitu reviewed (commented):
Added an MRE in the failing `test_file_metadata_timestamp_updated` test

tareknaser commented:
> Added an MRE in the failing test_file_metadata_timestamp_updated test

It looks like this issue might be related to the file system's timestamp granularity. From what I've found, it seems to depend on the file system being used.

For Linux:
- `EXT3` supports timestamps with a precision up to one second.
- `EXT4` can support timestamps with a precision up to nanoseconds, but [not always](https://stackoverflow.com/a/76581457)
> I am not entirely sure about these numbers. I couldn't find a definitive resource to list them.

I've pushed a commit that adds a one-second delay after each file write to fix the tests. This seems to be enough to fix the tests. I also reviewed the Rust file system documentation and added a call to [`sync_all()`](https://doc.rust-lang.org/std/fs/struct.File.html#method.sync_all).

I’m wondering if users might run into similar issues with `fs-storage`. It might be worth adding a note about this in the `FileStorage`’s doc comment.

twitu commented:
It's unfortunate that the OS does not update the metadata fast enough for it to be useful for us. However, we can perhaps work around it by setting the modified time ourself. With this logic we can set the modified timestamp consistently and the tests pass without needing a sleep.

```rust
        fs::create_dir_all(parent_dir)?;
        let mut file = File::create(&self.path)?;
        file.write_all(serde_json::to_string_pretty(&self.data)?.as_bytes())?;
        file.flush()?;

        let new_timestamp = SystemTime::now();
        file.set_modified(new_timestamp)?;
        file.sync_all()?;
```

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -29,9 +29,16 @@ pub struct FileStorage<K, V>
 {
+    /// Label for logging
     label: String,
+    /// Path to underlying file where data is persisted
```
```suggestion
    /// Path to the underlying file where data is persisted
```

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -190,16 +251,16 @@ where
+        file.write_all(serde_json::to_string_pretty(&self.data)?.as_bytes())?;
+        file.flush()?;
+
+        let new_timestamp = SystemTime::now();
+        file.set_modified(new_timestamp)?;
+        file.sync_all()?;
```
Thanks. This is a cool solution!

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -293,40 +360,79 @@ mod tests {
+            .modified()
+            .unwrap();
+        file_storage.write_fs().unwrap();
+        std::thread::sleep(std::time::Duration::from_secs(1));
```
Since we now have a clever solution, can we remove all the delays from the tests?

tareknaser reviewed (approved):
Thanks for making all the requested changes. I've added a few minor comments. 

Once you rebase, we should be ready to merge.

tareknaser commented on fs-storage/src/file_storage.rs:
```diff
@@ -176,12 +152,97 @@ where
     }
 
-    /// Write the data to the storage file
+    /// Write the data to file
```
We should also mention the trick of manually modifying the timestamp and why we need it as a note in the doc comment of this method

kirillt commented:
Good job @twitu @Pushkarm029 @tareknaser, this new logic looks very robust!
